slack: Refactor Slack text formatters.

PieterCK opened the pull request:
This PR aims to make `render_block` and ` render_attachments` be reusable (primarily for the Slack importer)  and clean up any duplicative functions(`replace_links` & `replace_formattings`) in `slack_incoming`. 

---
- This means, moving `render_block` and ` render_attachments` from `slack_incoming` to `slack_message_conversion`. (Done)

- Make `render_block` and ` render_attachments` compatible with `convert_slack_formatting` instead of `replace_links` & `replace_formattings` from `slack_incoming` (Commit: e919327bb1b529deb3daa5d74825d1ce5d6cd933). 

  However, if one tries to use the current regex in `convert_slack_formatting` directly in `slack_incoming`, some tests will fail:
  <details>
  
  <summary>
  Running tests in slack_incoming with previous `convert_slack_formatting` regex:
  </summary>
  
  ```bash
  ======================================================================
  FAIL: test_attachment_fields (zerver.webhooks.slack_incoming.tests.SlackIncomingHookTests)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/srv/zulip/zerver/webhooks/slack_incoming/tests.py", line 215, in test_attachment_fields
      self.check_webhook(
    File "/srv/zulip/zerver/lib/test_classes.py", line 2448, in check_webhook
      self.assert_channel_message(
    File "/srv/zulip/zerver/lib/test_classes.py", line 2464, in assert_channel_message
      self.assertEqual(message.content, content)
    File "/srv/zulip/zerver/lib/test_classes.py", line 290, in assertEqual
      raise AssertionError(
  AssertionError: Actual and expected outputs do not match; showing diff.
    Build bla bla succeeded
    
  - *Requested by*: Some user
  + **Requested by**: Some user
  - *Duration*: 00:02:03
  + **Duration**: 00:02:03
  - *Build pipeline*: ConsumerAddressModule
  + **Build pipeline**: ConsumerAddressModule
  - *Title with null value*
  + **Title with null value**
  - *Title without value*
  + **Title without value**
    Value with null title
    Value without title
  
  
  ======================================================================
  FAIL: test_complicated (zerver.webhooks.slack_incoming.tests.SlackIncomingHookTests)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/srv/zulip/zerver/webhooks/slack_incoming/tests.py", line 171, in test_complicated
      self.check_webhook(
    File "/srv/zulip/zerver/lib/test_classes.py", line 2448, in check_webhook
      self.assert_channel_message(
    File "/srv/zulip/zerver/lib/test_classes.py", line 2464, in assert_channel_message
      self.assertEqual(message.content, content)
    File "/srv/zulip/zerver/lib/test_classes.py", line 290, in assertEqual
      raise AssertionError(
  AssertionError: Actual and expected outputs do not match; showing diff.
    ## Hello from TaskBot
    
    Hey there 👋 I'm TaskBot. I'm here to help you create and manage tasks in Slack.
    There are two ways to quickly create tasks:
    
  - *1️⃣ Use the `/task` command*. Type `/task` followed by a short description of your tasks and I'll ask for a due date (if applicable). Try it out by using the `/task` command in this channel.
  + **1️⃣ Use the `/task` command**. Type `/task` followed by a short description of your tasks and I'll ask for a due date (if applicable). Try it out by using the `/task` command in this channel.
    
  - *2️⃣ Use the *Create a Task* action.* If you want to create a task from a message, select `Create a Task` in a message's context menu. Try it out by selecting the *Create a Task* action for this message (shown below).
  + **2️⃣ Use the *Create a Task* action.** If you want to create a task from a message, select `Create a Task` in a message's context menu. Try it out by selecting the *Create a Task* action for this message (shown below).
    
    [image1](https://api.slack.com/img/blocks/bkb_template_images/onboardingComplex.jpg)
    
    ➕ To start tracking your team's tasks, **add me to a channel** and I'll introduce myself. I'm usually added to a team or project channel. Type `/invite @TaskBot` from the channel or pick a channel on the right.
    
    ----
    
    [cute cat](https://pbs.twimg.com/profile_images/625633822235693056/lNGUneLX_400x400.jpg)
    
    👀 View all tasks with `/task list`
    
    ❓Get help at any time with:
    - `/task help`, or
    - type **help** in a DM with me
  
  
  ======================================================================
  FAIL: test_message_formatting (zerver.webhooks.slack_incoming.tests.SlackIncomingHookTests)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/srv/zulip/zerver/webhooks/slack_incoming/tests.py", line 40, in test_message_formatting
      self.assert_channel_message(
    File "/srv/zulip/zerver/lib/test_classes.py", line 2464, in assert_channel_message
      self.assertEqual(message.content, content)
    File "/srv/zulip/zerver/lib/test_classes.py", line 290, in assertEqual
      raise AssertionError(
  AssertionError: Actual and expected outputs do not match; showing diff.
  - **foo** *bar*
  + **foo** **bar**
  
  
  ======================================================================
  FAIL: test_message_with_actions (zerver.webhooks.slack_incoming.tests.SlackIncomingHookTests)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/srv/zulip/zerver/webhooks/slack_incoming/tests.py", line 85, in test_message_with_actions
      self.check_webhook(
    File "/srv/zulip/zerver/lib/test_classes.py", line 2448, in check_webhook
      self.assert_channel_message(
    File "/srv/zulip/zerver/lib/test_classes.py", line 2464, in assert_channel_message
      self.assertEqual(message.content, content)
    File "/srv/zulip/zerver/lib/test_classes.py", line 290, in assertEqual
      raise AssertionError(
  AssertionError: Actual and expected outputs do not match; showing diff.
    Danny Torrence left the following *review* for your property:
    
    [Overlook Hotel](https://google.com) 
     :star: 
     Doors had too many axe holes, guest in room 237 was far too rowdy, whole place felt stuck in the 1920s.
    
    [Haunted hotel image](https://is5-ssl.mzstatic.com/image/thumb/Purple3/v4/d3/72/5c/d3725c8f-c642-5d69-1904-aa36e4297885/source/256x256bb.jpg)
    
  - *Average Rating*
  + **Average Rating**
    1.0
  
  
  ======================================================================
  FAIL: test_message_with_attachment (zerver.webhooks.slack_incoming.tests.SlackIncomingHookTests)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/srv/zulip/zerver/webhooks/slack_incoming/tests.py", line 135, in test_message_with_attachment
      self.check_webhook(
    File "/srv/zulip/zerver/lib/test_classes.py", line 2448, in check_webhook
      self.assert_channel_message(
    File "/srv/zulip/zerver/lib/test_classes.py", line 2464, in assert_channel_message
      self.assertEqual(message.content, content)
    File "/srv/zulip/zerver/lib/test_classes.py", line 290, in assertEqual
      raise AssertionError(
  AssertionError: Actual and expected outputs do not match; showing diff.
    ## [[FIRING:2] InstanceDown for api-server (env="prod", severity="critical")](https://alertmanager.local//#/alerts?receiver=default)
    
  - :chart_with_upwards_trend: **[Graph](http://generator.local/1)**   :notebook: *[Runbook](https://runbook.local/1)*
  + :chart_with_upwards_trend: **[Graph](http://generator.local/1)**   :notebook: **[Runbook](https://runbook.local/1)**
    
  - *Alert details*:
  + **Alert details**:
  - *Alert:* api-server down - `critical`
  + **Alert:** api-server down - `critical`
  - *Description:* api-server at 1.2.3.4:8080 couldn't be scraped *Details:*
  + **Description:** api-server at 1.2.3.4:8080 couldn't be scraped **Details:**
       • **alertname:** `InstanceDown`
       • **env:** `prod`
       • **instance:** `1.2.3.4:8080`
       • **job:** `api-server`
       • **severity:** `critical`
    
  - *Alert:* api-server down - `critical`
  + **Alert:** api-server down - `critical`
  - *Description:* api-server at 1.2.3.4:8081 couldn't be scraped *Details:*
  + **Description:** api-server at 1.2.3.4:8081 couldn't be scraped **Details:**
       • **alertname:** `InstanceDown`
       • **env:** `prod`
       • **instance:** `1.2.3.4:8081`
       • **job:** `api-server`
       • **severity:** `critical`
  
  
  ======================================================================
  FAIL: test_message_with_blocks (zerver.webhooks.slack_incoming.tests.SlackIncomingHookTests)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/srv/zulip/zerver/webhooks/slack_incoming/tests.py", line 104, in test_message_with_blocks
      self.check_webhook(
    File "/srv/zulip/zerver/lib/test_classes.py", line 2448, in check_webhook
      self.assert_channel_message(
    File "/srv/zulip/zerver/lib/test_classes.py", line 2464, in assert_channel_message
      self.assertEqual(message.content, content)
    File "/srv/zulip/zerver/lib/test_classes.py", line 290, in assertEqual
      raise AssertionError(
  AssertionError: Actual and expected outputs do not match; showing diff.
    Danny Torrence left the following review for your property:
    
    [Overlook Hotel](https://example.com) 
     :star: 
     Doors had too many axe holes, guest in room 237 was far too rowdy, whole place felt stuck in the 1920s.
    
    [Haunted hotel image](https://is5-ssl.mzstatic.com/image/thumb/Purple3/v4/d3/72/5c/d3725c8f-c642-5d69-1904-aa36e4297885/source/256x256bb.jpg)
    
  - *Average Rating*
  + **Average Rating**
    1.0
  
  
  ----------------------------------------------------------------------
  Ran 20 tests in 2.234s
  
  FAILED (failures=6)
  
  ```
  </details>
  
  From those failing tests we can conclude that to make the `convert_slack_formatting` be compatible, it must:
  - Be able to capture formatted string multiline formatted strings (e.g.," \n_abc_\n*foo*"). (Commit: 682cd08b9d8dc8749ce945a1d3720041aebdf3d1)
  - Be able to capture formatted strings containing non-ASCII characters like emoji (e.g., "* :+1:  ok*"). (Commit: aaf92144a42ff97f81e922b53f867ea8e20dcf38)
  - Be able to capture formatted strings separated exactly by one character (e.g., `*abc*d*efg`). (Commit: 74da577212af88817ac3d46824feb371ba9a7d53)

Additionally, this also refactors the other regex capture groups to be more coherent

---

Next:
- #32163

Part of https://github.com/zulip/zulip/pull/31311.
<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented on zerver/webhooks/slack_incoming/tests.py:
```diff
@@ -207,7 +206,7 @@ def test_attachment_fields(self) -> None:
 **Duration**: 00:02:03
 **Build pipeline**: ConsumerAddressModule
 **Title with null value**
-**Title without value**
```
This is purposefully wrong, it should have double asterisks for bold. This is the same as `unsupported_bold_test` and `unsupported_bold_test_2`. Once those are fixed, this will naturally fail and then we can revert it back to `**Title with null value**`

PieterCK commented on zerver/tests/fixtures/slack_message_conversion.json:
```diff
@@ -84,6 +94,16 @@
+    },
+    {
+      "name": "unsupported_bold_test",
+      "input": "*foo* *bar*",
```
We couldn't capture this bold string because the `SLACK_BOLD_REGEX` have a capture group that matches for a character before asterisk "*"  
```python
SLACK_BOLD_REGEX = r"""
                    (\n|^|[ -(]|[+-/]|[:-?]|\{|\[|\_|\||\^|~)
                    ...
```
In this case, both `**foo**` and `*bar*` shares that one character, so one bold string won't get matched because that starting/ending character was used to match the other string. 

There's probably a way to support this case, but I don't think the potential complexity would be worth it because in practice Slack export shouldn't produce this kind of string, if two words got bolded separately like this example, then when exported it will just turn up like this: `*foo bar*`.

PieterCK commented on zerver/webhooks/slack_incoming/tests.py:
```diff
@@ -24,7 +24,6 @@ def test_message_formatting(self) -> None:
         tests = [
             ("some *foo* word", "some **foo** word"),
             ("*foo*", "**foo**"),
-            ("*foo* *bar*", "**foo** **bar**"),
```
This is has been moved to `slack_message_conversion.json` fixture as `unsupported_bold_test`. The difference is, instead of space, `unsupported_bold_test` use `\n` which is the more probable scenario.

PieterCK commented on zerver/tests/fixtures/slack_message_conversion.json:
```diff
@@ -84,6 +99,11 @@
       "conversion_output": "*~~italic~~* and ~~*strike*~~"
+    },
+    {
+      "name": "unsupported_bold_test",
```
This will also fail using the old reformatter, Slack incoming webhook's regex was able to format this string properly, but fail at other tests here, the trade-off is detailed here #31336.

PieterCK commented on zerver/tests/fixtures/slack_message_conversion.json:
```diff
@@ -84,6 +99,11 @@
+    {
+      "name": "unsupported_bold_test",
+      "input": "*abc*\n*def*\n*ghi*\n*jkl*\n\n*ABC ABC*",
+      "conversion_output": "**abc**\n*def*\n**ghi**\n*jkl*\n\n**ABC ABC**"
```
The reformatter couldn't support converting this bold string because the SLACK_BOLD_REGEX have a capture group that matches for a character before asterisk "*"

SLACK_BOLD_REGEX = r"""
                    (\n|^|[ -(]|[+-/]|[:-?]|\{|\[|\_|\||\^|~)
                    ...

In this case, both `*abc*` and `*def*` shares one character(`\n`) in between them, so `*def*` won't get matched because that character was used to match `*abc*`.

There's probably a way to support this case, but I don't think the potential complexity would be worth it because in practice Slack export shouldn't produce this kind of string, if two words got bolded separately like this example, then when exported it will just turn up like this: *foo bar*.

PieterCK commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -35,25 +48,25 @@
-                    (^|[ -(]|[+-/]|[:-?]|\{|\[|\_|\||\^|~)
+                    (\n|^|[ -(]|[+-/]|[:-?]|\{|\[|\_|\||\^|~)
                     (\*)
-                        ([ -)+-~—]*)([ -)+-~]+)                   # any character
```
The `([ -)+-~]+)` capture group is removed because is isn't used (`convert_markdown_syntax` doesn't use `match.group(5)). This applies to all other regexes

PieterCK replied:
CMIIW if this is important

PieterCK replied:
CMIIW and this turns out to be important

alexmv replied:
This is not quite the right replacement.  The previous regex required at least one character within  the `*`s -- the new one does not.  This should be:
```suggestion
                        ([^*]+)                 # any character except *
```

My other comment is if you your commit message is a bulleted list of "key changes" then maybe each one should be its own commit. :)

PieterCK replied:
I did another look through on the two inner capture groups, and yeah still couldn't figure out why there are two of them... Confirmed that only using the second inner capture groups would suffice.

I added this commit [slack_regex: Remove unnecessary inner capture group.](https://github.com/zulip/zulip/pull/32164/commits/a254abdac1871d81f7acc0137f9844467064e1d9) to remove the first capture groups before we rework the inner capture group in the commit after that.

PieterCK commented:
@laurynmm I think this is ready for review

laurynmm reviewed (commented):
@PieterCK - Thanks for splitting these changes out, but I think it's still hard to understand everything going on here with the current commit structure.

For example, it seems like both `render_block` and `render_attachment` are just being moved to another file by these changes, but still only being used in one place. Is that correct? I think that the idea of moving these functions was so that they would be used both for the incoming webhook and the import message conversion.

But maybe I'm missing something about what those changes are doing here?

Also, it looks like the changes to `replace_links` could be pulled out as a separate commit or even a small prep PR as those changes don't introduce any regressions/changes in how those are processed?

Also, the creation of `convert_slack_workspace_mentions` seems like it could be pulled out into it's own commit again because it doesn't impact the bold/italic/strike-through formatting changes.

In short, I think pulling all of those types of changes out (either into separate commits and/or a prep PR) will make it easier to understand and evaluate the changes that do alter how the bold/italic formatting is processed for Slack messages. Does that make sense?

Let me know if you have questions about anything above here or on CZO!

timabbott commented:
I merged the first commit -- moving that code to the new file seems fine. The second commit I consider unreviewable and needs to be split further, as Lauryn suggests. For example, one commit could just rename a function, another could change the bold/strikethrough/italics formatting (without any other changes), another could rework `replace_links`, etc. -- and each of those could be a nice readable commit with tests, a commit message explaining what it's aiming to achieve (ideally self-documented with tests for the cases where the new logic is an improvement). The goal is that it should feel really easy to verify that each commit is correct and an improvement. If you have all these different types of changes together in a single commit, it's really hard to verify the work.

This approach to making reviewable commits is discussed extensively in the Zulip commit guidelines.

PieterCK commented:
Thank you both for checking this out! Yeah the commit arrangements could definitely be better :sweat_smile: . This should be ready for review again by tomorrow :+1:

PieterCK commented:
Quite happy with how this turned out. I managed to work out the regex for the Slack text so there’s no trade-off or regression, and it’s still compatible with render_blocks and render_attachments, so I’m closing #31336. Let me know if any of the commits need to be squashed or split further! 

Tagging this as ready for another round of review, @laurynmm.

PieterCK commented:
Rephrased the commit descriptions and updated the commits referenced in the description.

laurynmm commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -79,52 +79,42 @@ def get_user_full_name(user: ZerverFieldsT) -> str:
-    # Map Slack's mention all: '<!channel>' to '@**all** '
-    # Map Slack's mention all: '<!here>' to '@**all** '
-    # No regex for this as it can be present anywhere in the sentence
+def convert_slack_workspace_mentions(text: str) -> str:
```
I think this commit could still be broken down into parts so that it's much easier to review.

Here's how I'd structure this in three commits:
1. Create and use the `convert_slack_workspace_mentions` helper. Have it be above `convert_to_zulip_markdown` in the file.
2. Move `convert_markdown_syntax` to where you want it in the file - before `convert_to_zulip_markdown` and after the helper you added in the previous commit.
3. Create and use the `convert_slack_formatting` helper. Again, position it in the file where you want it (before `convert_to_zulip_markdown` and after the function you moved in the last commit.

I know it may seem overly simplified, but that would allow me as a reviewer to look at each commit, verify what it does and it's correctness very quickly.

PieterCK replied:
Got it, other than the 3 commits you've suggested, I ended up splitting out the changes to reorder existing functions too, those starts with `slack_import: Rearrange  ...`

laurynmm commented on zerver/tests/fixtures/slack_message_conversion.json:
```diff
@@ -84,6 +84,11 @@
+    {
+      "name": "new_line_test",
+      "input": "\n*abc*\n",
+      "conversion_output": "\n**abc**\n"
```
I might be more explicit in the commit message that this output would have been `"\n*abc*\n"` prior to these changes.

PieterCK replied:
hmm here's how I've rephrased the commit description, does that look clearer?
```
Previously, the regex in `slack_message_conversion.py` couldn't capture
and reformat strings surrounded by newlines, such as "\n*ABC*\n". This
kind of string is likely not uncommon to be generated by both the Slack
exporter itself or by our Slack message block and attachment formatters
(`render_blocks` and `render_attachments`).
```

laurynmm commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -347,3 +347,9 @@ def render_attachment(attachment: WildValue) -> str:
     return "\n\n".join(piece.strip() for piece in pieces if piece.strip() != "")
+
+
+def replace_links(text: str) -> str:
```
I think this can be a separate commit too, which is done before all the bold/italic/strike-through regex changes. Just so that it's clear that this change doesn't depend on those updates.

PieterCK replied:
Split this out to : 49eca01cfc9245d3a4b2a2443a7d2179d059bbbf

alexmv replied:
I find adding a function which isn't used yet to be a weird single commit; I'd move this change to where it's necessary (the last commit of the series)

laurynmm reviewed (commented):
@PieterCK - Thanks for breaking these changes down into smaller commits! My main feedback is that they could be broken down a bit more so that the changes are still even more clear and easy to review for correctness.

Let me know if you have any questions about my feedback here or on CZO.

alexmv commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -49,22 +49,22 @@
-                             (\~)                                  # followed by an asterisk
+                             (\~)                                  # followed by an ~
+                                ([^~]*)            # any character except ~
+                             (\~)                                  # followed by an ~
```
```
Previously, the regex used multiple capture groups designed to
include everything except the formatting prefix and suffix
(`*`, `~`, `_`). This approach is replaced with a single capture
group that explicitly excludes these formatting characters,
simplifying the logic.
```

I don't think this description is accurate.  We still have different capture groups for the prefix and suffix after this change -- and we _need_ those to make it such that we can easily use the same replacement algorithm for any of these regexes once they mach.  The difference is that we drop a capture group for the last character inside those, which I'm really unclear on why we had in the first place (see below).

I think I would phrase the commit message as being about "not being ASCII-centric" -- the previous definition tried to match ranges of ASCII characters with a gap for `~`, and in so doing missed things outside the ASCII range.

There's something happening here that I don't quite grok, though -- the current regex prohibits a `-` from being the last character in the range.  Do we know why, or is that bogus and intentionally dropped in this commit?

alexmv replied:
The first sentence in the commit message also doesn't parse for me:
```
This prep commit to make `slack_incoming` use the same Slack
reformatting functions as the Slack import.
```

Doubly confusing because this _is_ the Slack import?  Which functions is it making these equivalent to?

PieterCK replied:
I've updated the commit message.

PieterCK replied:
>There's something happening here that I don't quite grok, though -- the current regex prohibits a - from being the last character in the range. Do we know why, or is that bogus and intentionally dropped in this commit?

I also not sure what's the purpose of that capture group (originally match group 5).

alexmv commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -163,13 +174,7 @@ def convert_to_zulip_markdown(
-    text = text.replace("<!everyone>", "@**all**")
-    text = text.replace("<!channel>", "@**all**")
-    text = text.replace("<!here>", "@**all**")
+    text = convert_slack_workspace_mentions(text)
```
First line on the commit message is a few characters too long right now -- maybe swap "Create and use the" to "Refactor a"

alexmv commented on zerver/tests/fixtures/slack_message_conversion.json:
```diff
@@ -94,6 +94,11 @@
       "conversion_output": "**1️⃣ bold** some *1️⃣ italic* word ~~1️⃣ strike~~"
+    },
+    {
+      "name": "overlapping_capture_group_test",
+      "input": "*abc*\n*def*\n*ghi*\n*jkl*\n\n*ABC ABC*",
+      "conversion_output": "**abc**\n**def**\n**ghi**\n**jkl**\n\n**ABC ABC**"
```
If this is the only overlapping case we care about, using `re.MULTILINE` instead of look-ahead is sufficient, since then we're still matching against `^` and `$` which are zero-width.  Are there examples _without_ `\n` being involved, which we could use?

PieterCK replied:
yes, I believe strings like this would produce the same scenario: `**Hello**!**How are you**?`

PieterCK replied:
hmm, I tried running this with the existing regex and `re.MULTILINE` as another flag and it still failed:
```bash
======================================================================
FAIL: test_message_conversion_fixtures (zerver.tests.test_slack_message_conversion.SlackMessageConversion) (slack_message_conversion='overlapping_capture_group_test')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/srv/zulip/zerver/tests/test_slack_message_conversion.py", line 53, in test_message_conversion_fixtures
    self.assertEqual(converted_text, test["conversion_output"])
  File "/srv/zulip/zerver/tests/test_slack_message_conversion.py", line 21, in assertEqual
    raise AssertionError(
AssertionError: Actual and expected outputs do not match; showing diff.
- *abc*
+ **abc**
- *def*
+ **def**
- *ghi*
+ **ghi**
- *jkl*
+ **jkl**
  
- *ABC ABC*
+ **ABC ABC**


----------------------------------------------------------------------
Ran 3 tests in 0.028s

```

```python
def convert_link_format(text: str) -> tuple[str, bool]:
    """
    1. Converts '<https://foo.com>' to 'https://foo.com'
    2. Converts '<https://foo.com|foo>' to '[foo](https://foo.com)'
    """
    has_link = False
    for match in re.finditer(LINK_REGEX, text, re.VERBOSE | re.MULTILINE):
        slack_url = match.group(0)
        url_parts = slack_url[1:-1].split("|", maxsplit=1)
        # Check if there's a pipe with text after it
        if len(url_parts) == 2:
            converted_url = f"[{url_parts[1]}]({url_parts[0]})"
        else:
            converted_url = url_parts[0]

        has_link = True
        text = text.replace(slack_url, converted_url)
    return text, has_link
```
it'd still fail using the new regex capture groups without the look-ahead matching...

PieterCK replied:
(Deleted a comment because it was incorrect.)

alexmv commented on zerver/webhooks/slack_incoming/view.py:
```diff
@@ -90,19 +95,6 @@ def api_slack_incoming_webhook(
 
     if body != "":
-        body = replace_formatting(replace_links(body).strip())
+        body = convert_slack_formatting(replace_links(body).strip())
```
I don't think I follow what the commit message means by "because of the extra formatting from `render_attachment` and `render_block`"?

PieterCK replied:
Concretely, formatted strings with emojis and newlines, where the strings are separated by exactly one character. But, I suppose those aren't exclusively produced by `render_attacment` nor `render_block`. I'll rephrase the commit description.

alexmv reviewed (commented):
I just gave this a once-over -- thanks for working to get these two disparate parts working together nicely!

PieterCK commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -48,25 +48,25 @@
 # Hence, ~stri~ke doesn't format the word in Slack, but ~~stri~~ke
 # formats the word in Zulip
 SLACK_STRIKETHROUGH_REGEX = r"""
```
hmm, I played around with the `re.M` option I found that although it can match strings like this:
```
*asbdasdas*
*asbdasdas*
*asbdasdas*
```
It's not enough to match this kind of newline:
```
\n*asbdasdas*\n*asbdasdas*\n*asbdasdas*\n
```
which can be produced by Slack's exporter:
![image](https://github.com/user-attachments/assets/11e49f7d-e675-4d42-8225-596765bc6cad)

PieterCK replied:
(Deleted previous reply because it is inaccurate)

PieterCK replied:
Hmm, so adding only the `re.MULTILINE` flag seems to not be enough. It can match strings like this:
```json
    {
      "name": "new_line_test",
      "input": "\n*abc*\n_helo_\n~stike~\n",
      "conversion_output": "\n**abc**\n*helo*\n~~stike~~\n"
    },
```

But somehow it'll reformat some of the string from this test multiple times :
```json
    {
      "name": "overlapping_bold_new_line_test",
      "input": "*alert:*\n*aescription:*\n*alert:*\n*description:*",
      "conversion_output": "**alert:**\n**aescription:**\n**alert:**\n**description:**"
    },
```
test result with only the `re.M` flag:
```
FAIL: test_message_conversion_fixtures (zerver.tests.test_slack_message_conversion.SlackMessageConversion) (slack_message_conversion='overlapping_bold_new_line_test')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/srv/zulip/zerver/tests/test_slack_message_conversion.py", line 52, in test_message_conversion_fixtures
    self.assertEqual(converted_text, test["conversion_output"])
  File "/srv/zulip/zerver/tests/test_slack_message_conversion.py", line 20, in assertEqual
    raise AssertionError(
AssertionError: Actual and expected outputs do not match; showing diff.
- ***alert:***
+ **alert:**
  **aescription:**
- ***alert:***
+ **alert:**
  **description:**

``` 

I haven't figured out why that is, but it can be resolved by adding the "\n" character to the first capture group.

alexmv replied:
Oh -- this is because we're doing the replacement wrong.  Specifically, we're using `text.replace` which replaces _all_ occurrences.  So with `*alert:*\n*alert:*` we find and match the first `*alert:*`, and replace *all occurrences* with `**alert:**`, generating `**alert:**\n**alert:**`.  But `finditer` is iterating over the matches to the _original_ value of `text`, so it come up with the second `*alert:*` match (despite that it wouldn't match if we ran it after doing that first replacement!).  So we again replace all `*alert:*` with `**alert:**` and get `***alert:***\n***alert:***`.

Adding `\n` to the regex us just papering over the problem -- since it's in the regex before the `^`, it means that the second match is `\n*alert:*` which _is no longer_ inside `**alert:**\n**alert:**` and so the `text.replace` does nothing.

The solution is to use the offsets of the regex's match to adjust the specific text that matched -- instead of `text.replace`:
```python3
# Map italic, bold and strikethrough Markdown
def convert_markdown_syntax(text: str, regex: str, zulip_keyword: str) -> str:
    """
    Returns:
    1. For strikethrough formatting: This maps Slack's '~strike~' to Zulip's '~~strike~~'
    2. For bold formatting: This maps Slack's '*bold*' to Zulip's '**bold**'
    3. For italic formatting: This maps Slack's '_italic_' to Zulip's '*italic*'
    """
    adjust_offset = 0
    for match in re.finditer(regex, text, re.VERBOSE | re.MULTILINE):
        converted_token = (
            match.group(1)
            + zulip_keyword
            + match.group(3)
            + match.group(4)
            + zulip_keyword
            + match.group(6)
        )
        text = text[:match.start(0)+adjust_offset] + converted_token + text[match.end(0)+adjust_offset:]
        adjust_offset += len(converted_token) - len(match.group(0))
    return text
```

I found this partially by looking at the replacement more carefully, and verified with `print()` statements inside `convert_markdown_syntax` to verify what eas happening at each replacement step when running the new test you added.

alexmv replied:
I'd suggest a new commit to fix the `text.replace` use, and another one after that adding `MULTILINE`.  The badness of using `text.replace` is only triggerable when the first/last group are zero width, which means it's not actually a visible bug until we introduce `MULTILINE`, but it's still misleading behaviour (e.g. `*foo* or *foo* or *foo*` goes through the `finditer` 3 times, but only the first time does it make any changes to `text`).

PieterCK replied:
Ah, ok that makes sense. Thanks for explaining how you debug that! 

If that's the case, I think `re.sub` might be suitable for this instead of offsetting the regex match:

```python
def convert_markdown_syntax(text: str, regex: str, zulip_keyword: str) -> str:
    """
    Returns:
    1. For strikethrough formatting: This maps Slack's '~strike~' to Zulip's '~~strike~~'
    2. For bold formatting: This maps Slack's '*bold*' to Zulip's '**bold**'
    3. For italic formatting: This maps Slack's '_italic_' to Zulip's '*italic*'
    """

    def replace_slack_format(match: Match[str]) -> str:
        return (
            match.group(1)
            + zulip_keyword
            + match.group(3)
            + match.group(4)
            + zulip_keyword
            + match.group(6)
        )

    return re.sub(regex, replace_slack_format, text, flags=re.VERBOSE)
```

I only did a limited amount of testing for this but it seems to do the job. I'll do a more thorough testing tomorrow and address your other feedbacks, but posting this now just in case you've gone down this route and found a pitfall.

PieterCK replied:
Ok, the `re.sub` approach I think does the same thing but a bit more concise, I separated the change to avoid replacing all identical matches to this commit: [slack_regex: Avoid replacing all identical matches.](https://github.com/zulip/zulip/pull/32164/commits/852bc99cc16c71219222587ea1ff087ae24f8d72)

PieterCK commented:
Thanks for reviewing this! I'll drop another comment once this is ready for another round of review

PieterCK commented:
Summary of each commits:
- **slack_regex: Update Slack regex handle multiline strings.**
  - I added the `re.M` flag.
  - The first capture groups still have the "\n" character because of [this reason](https://github.com/zulip/zulip/pull/32164#discussion_r1873030274)
  - Rewrote the commit description to reflect these changes.
- **slack_regex: Refactor Slack regex main capture group.**
  - Rewrote the commit description to address this [comment](https://github.com/zulip/zulip/pull/32164#discussion_r1870210879)
- **slack_regex: Fix overlapping capture groups.**
  - Adjusted the new `overlapping_capture_group_test` test fixture. See this [comment](https://github.com/zulip/zulip/pull/32164#discussion_r1870228578)
  - Rewrote the commit description to address this [comment](https://github.com/zulip/zulip/pull/32164#discussion_r1870204955)
-  **slack_incoming: Use Slack text reformatters from data import.**
    - Moved the change to refactor `replace_links` to this commit to address this [comment](https://github.com/zulip/zulip/pull/32164#discussion_r1870212169)
    - Rewrote the commit description to address this [comment](https://github.com/zulip/zulip/pull/32164#discussion_r1870234510).

PieterCK commented:
Some outstanding issues I'd like to highlight are:
- The purpose of the inner capture group is still unknown. https://github.com/zulip/zulip/pull/32164#discussion_r1870210879
- We still need "\n" in the first capture group to handle `overlapping_bold_new_line_test` but I don't know the reason.https://github.com/zulip/zulip/pull/32164#discussion_r1873030274

laurynmm commented:
Also, noting here that the first three prep commits here have been separated out into a separate PR, #32430, that I've reviewed and looks good to merge.

PieterCK commented:
Rebased for #32430 merge and updated the PR description

timabbott commented:
Cool, @alexmv is probably the best person to review this from here.

alexmv commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -52,21 +52,21 @@
                                  ([^~]*)              # any character except asterisk
                              (\~)                                  # followed by an asterisk
-                             ($|[ -']|[+-/]|[:-?]|\*|\_|\}|\)|\]|\||\^)  # ends with specified characters
+                             (?=$|[ -']|[+-/]|[:-?]|\*|\_|\}|\)|\]|\||\^)  # ends with specified characters
```
This commit message describes the problem, but does not contain any part which descibes what the fix is, or why it is correct.

However, the example (and test) that it shows isn't even correct -- the new test passes even without this commit. Because we match bold and italic in different passes, the whitespace between those words isn't a problem.  `*foo* *bar* *baz*` is a problem, but not `*foo* _bar_ ~baz~`.

Here's what I'd write:
```
slack_regex: Fix overlapping capture groups.

The Slack text regexes match specific characters before and after the
formatted string to ensure that they only match at word boundaries.

However, because the capture groups consume each matching character,
including the characters used to determine word boundaries, two
formatted strings separated by a single matching character result in one
string not being matched, as the trailing whitespace is already
consumed and cannot also match as the leading pre-match whitespace for
the next character.

Switch to a look-ahead regex for the trailing word boundary
characters.  This is zero-width, and as such the next match can still
also consume the same characters.
```

PieterCK replied:
Ok! I've updated the commit description as per suggested and also the test fixture to:
```json
    {
      "name": "overlapping_capture_group_test",
      "input": "*foo* *baz* *bar*",
      "conversion_output": "**foo** **baz** **bar**"
    }
```

PieterCK commented:
@alexmv - I've addressed your comments. I appreciate the explanations and suggestions! especially on the blocker I was having trouble with :sweat_smile:.

Summary of changes:
- I added [slack_regex: Avoid replacing all identical matches.](https://github.com/zulip/zulip/pull/32164/commits/ec09d7619c8783f7b087d7b3ce18525fc6880216) to address [this](https://github.com/zulip/zulip/pull/32164/files#r1904432231). I used a different approach of using `re.sub` than [suggested](https://github.com/zulip/zulip/pull/32164/files#r1904418461) because of [this](https://github.com/zulip/zulip/pull/32164/files#r1916710691).
- I added [slack_regex: Remove unnecessary inner capture group.](https://github.com/zulip/zulip/pull/32164/commits/a254abdac1871d81f7acc0137f9844467064e1d9) to address [this](https://github.com/zulip/zulip/pull/32164/files#r1904435304).
- in [slack_regex: Simplify Slack regex main capture group.](https://github.com/zulip/zulip/pull/32164/commits/f3c45370eb7d578a04203c6ad3e547fa8ea68b49) I changed the inner capture group to capture one or more matches as per [suggested](https://github.com/zulip/zulip/pull/32164/files#r1904435304).
- In [slack_regex: Fix overlapping capture groups.](https://github.com/zulip/zulip/pull/32164/commits/4661b8c67079632ad293a92bf3617da99874e427):
  - I fixed the new test, which was inaccurate in showing the problem with overlapping capture groups.
  - Rewrote the commit description as per [suggested](https://github.com/zulip/zulip/pull/32164/commits/4661b8c67079632ad293a92bf3617da99874e427#r1904446533).
- I added [integrations: Adjust Slack formatting tests.](https://github.com/zulip/zulip/pull/32164/commits/d2096a9859488c99e1cff557899b1a36a8d4e226) it turned out that these changes have fixed some of the problematic Slack text documented there too :smiley:

Other than that, I rewrote the other commits descriptions to reflect the changes mentioned above, such as moving a change out to its own commit.

I think this is clear of any blocker and is ready for another look-through! Let me know if I've missed anything

alexmv commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -51,21 +51,21 @@
 SLACK_STRIKETHROUGH_REGEX = r"""
                              (^|[ -(]|[+-/]|\*|\_|[:-?]|\{|\[|\||\^)     # Start after specified characters
```
The same theory applies to the absolute jumble that is the "start after specified characters" mess.  This regex is a range which includes every symbol character except the closing delimiters (`)}]`), the tilde (`~`), the backtick (`` ` ``), the backslash (for some reason?), and the `@`.

I think we should express that in a way which is Unicode-aware and more coherent:
```
(?![~`@\\\p{Pf}\p{Pe}])[\p{P}\p{Zs}\p{S}]
```

That's "except for `` ~`@\ `` and closing quotes or parens, look for punctuation, whitespace, or symbols."   `(?! ... )` is a negative look-ahead assertion, saying "without actually consuming any characters, assert that this _must not match here_).  You can compare how this version works for the entire printable ASCII range: [old](https://regex101.com/r/VKkmJs/1), [new](https://regex101.com/r/HFSUUD/2).  But it's also going to work better for folks using, e.g., «quotes», and is much more explicit about what it's _not_ matching without memorizing the ASCII character chart.

You should be able to apply that same fix, modulo just a different one exception character, to lines 59 and 66 -- and an adjusted version that skips _opening_ versions (using `\p{Pi}\p{Ps}` instead of `\p{Pf}\p{Pe}`) for lines 56, 63, and 70.

This cleanup should all be in its own commit.  The "Start after specified characters" comment is also ... a totally useless comment at present, and should either be removed or improved to add value.

PieterCK replied:
Alright, the new pattern looks great, it only missed the pattern for start of line `^`, I think. I added that in the very last bit of the pattern.

- Added `slack_regex: Use the regex module`. because we don't use the `regex` module for this, and the `re` module doesn't support Unicode pattern like `\p{Pf}`.
- Added `slack_regex: Refactor the first capture groups` for changes to the first capture groups.
- Added `slack_regex: Refactor the last capture groups` for changes to the last capture groups. I've also added a test for Unicode:
  ```json
      {
        "name": "unicode_quotes",
        "input": "«~strike~» 「*bold*」 ❰_italic_❱",
        "conversion_output": "«~~strike~~» 「**bold**」 ❰*italic*❱"
      }
  ```
- In ` slack_regex: Fix overlapping capture groups`, I updated how the last capture group looks ahead. Previously it was: 
  ```
  (?=$|[ -']|[+-/]|[:-?]|\}|\)|\]|\_|\||\^|~)
  ```

PieterCK replied:
I dug around and was able to find a discussion on the original pattern. Thread: https://github.com/zulip/zulip/pull/7933#discussion_r159731146.

One thing I'd like to highlight is why the current pattern avoids mid-word patterns like (`test~t~`). The reasoning was that Slack doesn't capture that kind of string. I manually tested out the theory, and indeed one can't _manually_ format mid-word string, but using the format button, it's possible:
![image](https://github.com/user-attachments/assets/f6dc0020-7b12-4c89-b814-c4d922f6e820)

Considering mid-word formatting is also possible in Zulip, I wonder if we still want to preserve this rule—excluding alphanumeric characters in the first capture group—for the new pattern.

alexmv commented on zerver/webhooks/slack/tests.py:
```diff
@@ -271,7 +271,7 @@ def test_message_with_code_block(self) -> None:
 
     def test_message_with_complex_formatted_texts(self) -> None:
-        message_body = "this is text messages with overlapping formatting\n_**bold with italic**_\n~**bold with strike through**~\n~*italic with strike through*~\n~***all three***~"
+        message_body = "this is text messages with overlapping formatting\n***bold with italic***\n~~**bold with strike through**~~\n~~*italic with strike through*~~\n~~***all three***~~"
```
These test changes should happen in the same commit that caused them to no longer pass as written.  Tests should be passing at every commit in a series.

PieterCK replied:
Arg.., right! Not sure why I separated these changes

PieterCK replied:
Previously commit `integrations: Adjust Slack formatting tests.` fixed a couple test in Slack webhook integration, I moved the changes in that commit to the earlier commit where they were introduced: 
- Fixed `test_message_with_complex_formatted_texts` in commit: `slack_regex: Update Slack regex handle multiline strings`
- Fixed `test_message_with_complex_formatted_mentions` in commit: `slack_regex: Fix overlapping capture groups.`

PieterCK commented:
@alexmv - Thanks for review again. I've detailed the changes to fix the Slack webhook integration tests in [here](https://github.com/zulip/zulip/pull/32164#discussion_r2046072362), and the changes for the first and last capture groups, [here](https://github.com/zulip/zulip/pull/32164#discussion_r2050510347). This is ready for you to take a look again,

PieterCK commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -51,21 +51,21 @@
                              (^|[ -(]|[+-/]|\*|\_|[:-?]|\{|\[|\||\^)     # Start after specified characters
                              (\~)                                  # followed by an asterisk
-                                 ([ -)+-}—]*)([ -}]+)              # any character except asterisk
+                                 ([ -}]+)              # any character except asterisk
```
lines re-aligned :+1:

alexmv commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -62,7 +62,15 @@
                                  ([ -)+-}—]*)([ -}]+)              # any character except ~
                              (\~)                                  # followed by a ~
-                             ($|[ -']|[+-/]|[:-?]|\*|\_|\}|\)|\]|\||\^)  # ends with specified characters
+                            (
```
This is missing a space to line up.

PieterCK replied:
Fixed

alexmv commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -62,7 +62,15 @@
+                                # line.
+                                # Skip ~ to not reformat the same string twice
+                                # Skip @ and \
+                                # Skip opening brackets & closing quote (\p{Pi}\p{Ps})
```
```suggestion
                                # Skip opening brackets & opening quote (\p{Pi}\p{Ps})
```

PieterCK replied:
Done!

alexmv reviewed (commented):
Only minor nits!

I don't think the opening and closing capture group changes need to be two separate commits, and I would also merge the swap to `regex` into that change, so it's bundled with the changes it's relevant to.

PieterCK commented:
Great! I've squashed the 3 commits into: `slack_regex: Refactor how the capture groups are defined.`

alexmv commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -48,25 +50,57 @@
                              """
 SLACK_ITALIC_REGEX = r"""
-                      (^|[ -*]|[+-/]|[:-?]|\{|\[|\||\^|~)
+                    # Same as `SLACK_STRIKETHROUGH_REGEX`s. The difference
+                    # being, this skips _ instead of ~
```
This block wants two more spaces to line up its indentation.

timabbott commented:
Fixed the small indentation issue so I could merge this, thanks @PieterCK and @alexmv!
